### PR TITLE
[libc++] Add assertions for potential OOB reads in std::nth_element

### PR DESCRIPTION
**Pull request by danlark1**
Same as https://reviews.llvm.org/D147089 but for std::nth_element

### CODE REVIEW

**Comment by danlark1** [comment]
I have a fun error in debug mode

```
/home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-e33075b085b1-1/llvm-project/libcxx-ci/build/generic-debug-mode/include/c++/v1/__algorithm/nth_element.h:46:1: error: function '__nth_element' has cognitive complexity of 144 (threshold 143) [readability-function-cognitive-complexity,-warnings-as-errors]
```

**Comment by danlark1** [comment]
Ping

File: libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp:55]:
```suggestion
    explicit ComparisonResults(std::string_view data) {
```

File: libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp:68]:
```suggestion
    bool compare(size_t* left, size_t* right) const {
```

File: libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp:178]: I can't comment on it due to the Github review UI, but there's a line above in this file with:

```
std::nth_element(copy.begin(), copy.end(), copy.end(), checked_predicate); // doesn't go OOB even with invalid comparator
```

I think we should remove that part of the test above since clearly it's not that `nth_element` was not affected, it's just that the specific values I used didn't trigger the issue.

File: libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator.pass.cpp:1]: As I discussed in https://reviews.llvm.org/D150264#inline-1504645, I am a bit uncomfortable with the fact that this test checks both for OOB accesses and for the `_LIBCPP_DEBUG_STRICT_WEAK_ORDERING_CHECK` diagnostics. We should really have a test that checks for these OOB accesses and that is completely separate from the strict-weak-ordering check, which is a debug-mode only check.

Not for this patch, but a follow-up to split this up into two tests would be really great.

**Review by ldionne — APPROVED** [review]
Thanks a lot for the patch! I have a few comments but this is basically good. You can squash+merge once the review comments are addressed and the CI is green.

**Comment by danlark1** [comment]
> Thanks a lot for the patch! I have a few comments but this is basically good. You can squash+merge once the review comments are addressed and the CI is green.

Please, merge for me, I don't have commit rights

**Comment by ldionne** [comment]
I'll wait for the CI to be green and then I'll merge. I wish we had auto-merge enabled in the repo.

**Comment by danlark1** [comment]
@ldionne, CI is almost green, the errors are in the format of

```
error: function '__nth_element' has cognitive complexity of 144 (threshold 143) [readability-function-cognitive-complexity,-warnings-as-errors]
```

What should I do about it?

**Comment by ldionne** [comment]
@philnik777 These clang-tidy errors should really not be errors, do you have a preference for how to deal with those?

Also @danlark1 if you rebase onto `main` the issues with macOS in the CI should be fixed now.

**Comment by philnik777** [comment]
> @philnik777 These clang-tidy errors should really not be errors, do you have a preference for how to deal with those?

That depends on what we want to do with these kinds of issues in general. I think we should avoid making functions so complex. If we have a general consensus there, I think we should just suppress the warning with a todo that the function should be split up (or simplified some other way). If not, then I guess we can just disable the clang-tidy check. Having it non-fatal is somewhat useless IMO, just like having compiler warning non-fatal is useless.

**Comment by danlark1** [comment]
>Also @danlark1 if you rebase onto main the issues with macOS in the CI should be fixed now.

@ldionne, Done

**Comment by ldionne** [comment]
> That depends on what we want to do with these kinds of issues in general. I think we should avoid making functions so complex. If we have a general consensus there, I think we should just suppress the warning with a todo that the function should be split up (or simplified some other way).

I think I agree. I just looked at the function and if this were code being committed today, it would likely never pass code review in that state (I'm not judging what was done at that time, but realistically we would ask for the function to be broken up -- I think).

@philnik777 is there a way to turn off clang-tidy warnings selectively? Do we simply use the usual pragma to disable diagnostics?

**Comment by philnik777** [comment]
You can use `// NOLINT(check-name)` or `// NOLINTNEXTLINE(check-name)` to disable clang-tidy warnings. I would probably put a `// NOLINTNEXTLINE(readability-function-cognitive-complexity)` above the function.

**Comment by ldionne** [comment]
Thanks!

**Pull request merged**
